### DISCUSSION TITLE
WhichKey keybindings: what "style" to prefer?

### DISCUSSION THREAD

**Discussion by xeluxee** [LunarVim/LunarVim#1690] [Ideas]
Hi there,
I've used [SpaceVim](https://github.com/SpaceVim/SpaceVim) for a long time, but now I'm using LunarVim to take advantage of the new features introduced in neovim 0.5.
After switching one thing I'm missing are some useful `<space>` keybindings. Most of them are made by three keystrokes (`space` + `letter1` + `letter2`).
Instead I've noticed that most of LunarVim keybindings are made by two keystrokes (except for +LSP, +git and something else).

I'm wondering if this is related to typing speed: two-strokes keybindings are faster to type, but with three-strokes keybindings more things can be mapped.

Here only few examples from SpaceVim that I find useful:
`<space>ff` → find file (currently `<space>f`)
`<space>fs` → save file (currently `<space>w`)
`<space>ws` → split window horizontally
`<space>wv` → split window vertically
`<space>tb` → toggle background color (from light to dark and vice-versa)
`<space>ts` → toggle spell checking
`<space>xw` → remove trailing whitespaces

I know I can create my own keybindings in `config.lua` (and I've already done it), but I think that using three-strokes keybindings could be advantageous for some (or many) users.

So I'm asking you what's your opinion about this, and what keybindings style (two or three keystrokes) do you prefer.

**Comment by chase** [comment]
If I use certain behaviors constantly, I prefer two stroke. If I use it somewhat often, I use three strokes. Anything less often than that I'll just use a command.

**Reply by xeluxee** [reply]
I agree with that
I know that `<leader>w` is probably the most used whichkey keybinding, but what about replacing `<leader>w` with `+Windows` group?
I think having keybindings to manage windows in LunarVim by default could be useful
``` lua
lvim.builtin.which_key.mappings.w  = {
  name = "+Windows",
  p = {"<C-w>p",                     "Previous Window"},
  g = {"<cmd>ChooseWin<cr>",         "Go To Window"},
  c = {"<C-w>c",                     "Close Window"},
  d = {"<C-w>c",                     "Close Window"},
  h = {"<C-w>h",                     "Window Left"},
  j = {"<C-w>j",                     "Window Below"},
  l = {"<C-w>l",                     "Window Right"},
  k = {"<C-w>k",                     "Window Up"},
  r = {"<cmd>WinResizerStartResize<cr>", "Window Resizer"},
  m = {"<C-w>o",                     "Maximize window"},
  ["="] = {"<C-w>=",                 "Balance Window"},
  s = {"<C-w>s",                     "Split Window Horizontally"},
  v = {"<C-w>v",                     "Split Window Vertically"},
  ["-"] = {"<C-w>s",                 "Split Window Horizontally"},
  ["|"] = {"<C-w>v",                 "Split Window Vertically"},
}
```

**Reply by xeluxee** [reply]
And what about replacing `<leader>q` with this?
``` lua
lvim.builtin.which_key.mappings.q = {
  name = "+Quit",
  q = {"<cmd>qa<cr>",  "Quit Neovim"},
  w = {"<cmd>wqa<cr>", "Quit Saving Everything"},
  Q = {"<cmd>qa!<cr>", "Quit Anyway (without saving)"},
}
```

**Comment by ssbanerje** [comment]
Yeah, I was a long-time Spacevim user as well.

The keys that I miss the most are the `<Space>0-9` keys for jumping between windows. I had to figure out how put the window numbers in the statusline and then remap the keys. Still haven't figured out how to do the nvim-tree and symbols sidebars.

The other thing that I miss is documentation. Seems like very few of the Lua plugins have nice help tags like the viml plugins in SVIM. That definitely makes changing things much harder. It seems like people document the `setup()` function for the plugin and not a lot of the commands/mappings/APIs that get added. :disappointed:

**Reply by kylo252** [reply]
> Yeah, I was a long-time Spacevim user as well.
> 
> The keys that I miss the most are the `<Space>0-9` keys for jumping between windows. I had to figure out how put the window numbers in the statusline and then remap the keys. Still haven't figured out how to do the nvim-tree and symbols sidebars.

Barbar defintely supports that. Try `:BufferPick`

> The other thing that I miss is documentation. Seems like very few of the Lua plugins have nice help tags like the viml plugins in SVIM. That definitely makes changing things much harder. It seems like people document the `setup()` function for the plugin and not a lot of the commands/mappings/APIs that get added. 😞

I'm gonna make an effort to update master today, then we can finally update lunarvim.org and not worry about it having conflicts with what's used on master.

**Reply by ssbanerje** [reply]
No helptags on `BufferPick` :smiley: :sob:

**Reply by xeluxee** [reply]
> Still haven't figured out how to do the nvim-tree and symbols sidebars.

What do you mean? Are you referring to [tagbar](https://github.com/preservim/tagbar) with symbol siderbar?

**Reply by ssbanerje** [reply]
I have been using simrat39/symbols-outline.nvim...

**Reply by xeluxee** [reply]
Oh ok I was thinking to tagbar because it was used in SpaceVim :sweat_smile:
